Detectors/FIT/FT0/workflow/src/FT0DataReaderDPLSpec.cxx

```diff
@@ -0,0 +1,98 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// @file   FT0DataReaderDPLSpec.cxx
+
+#include "FT0Workflow/FT0DataReaderDPLSpec.h"
+
+using namespace o2::framework;
+
+namespace o2
+{
+namespace ft0
+{
+using namespace std;
+void FT0DataReaderDPLSpec::init(InitContext& ic)
+{
+}
+
+void FT0DataReaderDPLSpec::run(ProcessingContext& pc)
+{
+  auto size = pc.inputs().size();
+  LOG(INFO) << size;
+  DPLRawParser parser(pc.inputs());
+  mVectorData.clear();
+  uint64_t count = 0;
+  for (auto it = parser.begin(), end = parser.end(); it != end; ++it) {
+    count++;
+    auto rdhPtr = it.get_if<o2::header::RAWDataHeader>();
+    const unsigned char* dataSource = it.data();
+    const size_t dataSize = it.size();
+    /*
+    LOG(INFO) <<"---------------------------";
+    LOG(INFO) <<"size: "<<it.size();
+    LOG(INFO) <<"blockLength: "<<rdhPtr->blockLength;
+    LOG(INFO) <<"memorySize: "<<rdhPtr->memorySize;
+    LOG(INFO) <<"offsetToNext: "<<rdhPtr->offsetToNext;
+    LOG(INFO) <<"feeId: "<<rdhPtr->feeId;
+    LOG(INFO) <<"linkID: "<<unsigned(rdhPtr->linkID);
+    LOG(INFO) <<"heartbeatOrbit: "<<rdhPtr->heartbeatOrbit;
+    LOG(INFO) <<"---------------------------";
+    */
+    for (size_t iBytes = 0; iBytes < dataSize;) {
+      //Getting header
+      RawEventBlock rawEventBlockFT0;
+      EventHeaderFT0& eh = rawEventBlockFT0.mEventHeader;
+      eh.deserialize(dataSource + iBytes);
+      auto nGBTwords = eh.nGBTWords;
+      std::vector<EventDataFT0>& vectorEventData = rawEventBlockFT0.mVectorEventData;
+      iBytes += SIZE_WORD;
+      //checking nGBTwords
+      if (nGBTwords > 0 && nGBTwords < 7) {
+        for (int iWord = 0; iWord < nGBTwords && iBytes < dataSize; iWord++) {
+          //First 40-bit word
+          vectorEventData.emplace_back();
+          EventDataFT0& edFirst = vectorEventData.back();
+          edFirst.deserialize(dataSource + iBytes);
+          iBytes += CompressedEventData::PayloadSizeFirstWord;
+          //Second 40-bit word
+          vectorEventData.emplace_back();
+          EventDataFT0& edSecond = vectorEventData.back();
+          edSecond.deserialize(dataSource + iBytes);
+          iBytes += CompressedEventData::PayloadSizeSecondWord;
+        }
+      }
+      mVectorData.push_back(rawEventBlockFT0);
+    }
+  }
+  pc.outputs().snapshot(Output{o2::header::gDataOriginFT0, "RAWEVENTBLOCK", 0, Lifetime::Timeframe}, mVectorData);
```

Comment:
As far as I can see, you are simply converting a POD binary data to vector of vectors (should be avoided as a content of a message, non-POD) with exactly the same content but with larger size. What is the purpose of this conversion?
I thought, the decoding of raw data should lead to either more compact data (for CTF) or to digit-like POD data which can be fed into the reconstruction.

Reply:
Hello Ruben,
this code was created by me. Main task of this code is the payload de-serialization, and transferring through DPL channel to QC, or for some further operations on data(filtering, etc), which will be discussed in the future. 
Each raw data block contains 1 EventHeader and up to 12 EventData(one for each FT0 channel) objects.
So, I decided to prepare class(RawEventBlock) with this structs as fields(header + dynamic container for EventData). But such structs(EventHeader and EventData) are not POD objects(checking by std::is_pod returns false) and also they cannot be serialized as ROOT object through default Streamer ("error in <TStreamerInfo::Build>: o2::ft0::Event Header, unknown type: o2::ft0::EventHeader::(anonymous)"), because streamer doesn't know how to stream fields in such structs(structs containing unions, which combine array of uint64_t's and POD struct, it's too complicated for default streamer).

First solution was the defining of custom streamer, but I failed.
Second solution is temporary, just using simple POD structs(EventHeaderFT0 and EventDataFT0) in the RawEventBlock class.

About devices:
- "Reader" is a device for payload de-serialization and transferring through DPL.
- "Process" (Processor) is just for checking(dumping), if the data were de-serialized and transferred correctly.

I will prepare README.md and more comments for clarification.

If I understood correctly, I should use only container of EventDataFT0 with static size(span) inside RawDataBlock, and the same option for container of RawEventBlocks which will be transferred through DPL.
I will fix this.

Is it possible to solve problem with such "complicated" structs(EventHeader and EventData) as fields inside class, without creation of similar POD structs(EventHeaderFT0 and EventDataFT0) with the same functionality? Should I change my schematics and use another solution for this task?

Best regards,
Artur Furs.

Reply:
Hi @afurs

1st of all, the objects don't need to formal std::pod to be messaged by the DPL, it is enough to have them mesageable as
both EventHeader and EventData are:
````  
LOG(INFO) << "o2::ft0::EventHeader, is_messageable::value = " << o2::framework::is_messageable<o2::ft0::EventHeader>::value;
LOG(INFO) << "o2::ft0::EventData, is_messageable::value = "  << o2::framework::is_messageable<o2::ft0::EventData>::value;
...
[INFO] o2::ft0::EventHeader, is_messageable::value = 1
[INFO] o2::ft0::EventData, is_messageable::value = 1
````
which means that you can send a ``vector<EventHeader>`` or ``vector<EventData>`` w/o any serialization and receive them as spans (no copy). But obviously, you cannot do this for vector of vectors, which should be avoided.

You can mimic "dynamic" records by creating a vector<EventData> and another vector<EventHeaderAndRef> with
EventHeaderAndRef being a 
````
struct EventHeaderAndRef : public EventHeader
{
  RangeReference<int,int> ref;
}
````
with every element of EventHeaderAndRef containing the header + entry of the 1st EventData in the 1st vector + number of entries (ref.getFirstEntry() and ref.getEntries() resp.). It is the same mechanism which T0 uses to store/send the digits data, i.e. Digit.h and https://github.com/AliceO2Group/AliceO2/blob/dev/DataFormats/Detectors/FIT/FT0/include/DataFormatsFT0/ChannelData.h.

But what is the point of this "deserialization of raw data at all? What do you gain by inflating it in size by factor of few and still sending logically the same info, just rearranged from linear stream to vector of vectors?

Can't you directly convert it to digits / channels data?

Cheers,
 Ruben

Reply:
Hello @shahor02 ,
okay, I will apply these solutions, and I will avoid things you told.

Main goal of raw data deserialization is the simplification of extracting data from DPL channel, in required place(QC task, aggregator, etc ). Also, it seems that we will need to repack some FT0 raw data blocks, because there could be a large amount of them under given orbit, and some blocks from the previous orbit can be pushed to the page with next orbit.
If we have to achieve strong processor/memory optimization, I can prepare decoder/encoder and use them "in a given place". Only binary data will be transferred in this case. 
Another option is converting raw data blocks into digits, and transferring them through DPL channels. Is it allowed at FLP? Actually we are discussing with @AllaMaevskaya this option. 

Best Regards,
Artur Furs.

Reply:
Hi @afurs 
I did not really understand what this repacking means, could you add some details?
Apart from that, I don't understand what is the point if this deserialization, if as a result of it you inflate the data size by a factor of a few. As far as I understand, conversion or your raw data to digits does not require particular computations, what are pros some intermediate deserialized format wrt digits? Do you have some estimate of the output size (including QED contribution)?
You can do whatever you want on the FLP after the STFbuilder: you don't need to stick to RDH format, just should provide to STFsender a TF payload to send. Obviously, the payload should contain all necessary interaction records to make possible matching to other detectors.

Reply:
Hello @shahor02 ,
amount of event data blocks(EventHeader + up to 12 EventData) could exceed size of the last page in a given orbit, and FEE will push last blocks to the page with next orbit. In other words, there could be situation when EventHeader::orbit < RDH::orbit within page. It means that in such situation we will need to move event data block to the page/HBf with corresponding RDH::orbit.

But after this discussion, it seems that we need to deserialize raw data into data blocks(EventHeader+EventData, intermediate deserialized format) only for digit preparation within corresponding DPL and we should work only with digits. All actions, similar to "repacking",  will be done automatically during digit preparation. So, I will make global changes in this code. Estimating of the output size including QED is still in process.

Thanks a lot for remarks, now I have clear picture about this task!

Best Regards,
Artur Furs.

Reply:
OK for converting to digigs. For the data of given orbit written under the RDH of the next one: do you mean that reaching 8KB limit of the CRU page? Why don't you open a new page with the same RDH (just pageCounter incremented), as all other detectors do?

Reply:
Sorry, I was mistaken about incorrect conformity between EventHeader::orbit and RDH::orbit within page. My colleagues confirmed that it is impossible. Repacking means only collecting raw data from all links, i.e. preparing digits.